tests/test_georaster.py

```diff
@@ -467,22 +470,139 @@ def test_astype_uint8_to_int32_conversion():
+        [[0, 0], [0, 51]],
+        [[101, 153], [203, 255]]], dtype=np.uint8),
+        affine=some_float32_raster.affine, crs=some_float32_raster.crs, nodata=None)
+    assert raster_uint8 == expected_raster_uint8
+
+
+def test_astype_float32_to_uint8_conversion_without_stretching():
+    raster_float32 = make_test_raster(10.0, band_names=[1, 2], dtype=np.float32)
+    raster_uint8 = raster_float32.astype(np.uint8, in_range=None, out_range=None)
+    expected_raster_uint8 = make_test_raster(10, band_names=[1, 2], dtype=np.uint8)
+    assert raster_uint8 == expected_raster_uint8
+
+
+def test_astype_float32_to_int8_conversion():
+    with pytest.warns(GeoRaster2Warning):
+        raster_uint8 = some_float32_raster.astype(np.int8)
+    expected_raster_uint8 = GeoRaster2(image=np.array([
+        [[-128, -76], [-25, 25]],
+        [[50, 76], [101, 127]]], dtype=np.int8),
+        affine=some_float32_raster.affine, crs=some_float32_raster.crs, nodata=None)
+    assert raster_uint8 == expected_raster_uint8
+
+
+def test_astype_float32_to_int8_conversion_with_clip_negative():
+    with pytest.warns(GeoRaster2Warning):
+        raster_uint8 = some_float32_raster.astype(np.int8, clip_negative=True)
+    expected_raster_uint8 = GeoRaster2(image=np.array([
+        [[0, 25], [50, 76]],
+        [[88, 101], [114, 127]]], dtype=np.int8),
+        affine=some_float32_raster.affine, crs=some_float32_raster.crs, nodata=None)
+    assert raster_uint8 == expected_raster_uint8
+
+
+def test_astype_float32_to_float16_conversion():
+    with pytest.warns(GeoRaster2Warning):
+        raster_float16 = some_float32_raster.astype(np.float16, out_range=(0.0, 10.0))
+    expected_raster_float16 = GeoRaster2(image=np.array([
+        [[0.0, 2.0], [4.0, 6.0]],
+        [[7.0, 8.0], [9.0, 10.0]]], dtype=np.float16),
+        affine=some_float32_raster.affine, crs=some_float32_raster.crs, nodata=None)
+    assert raster_float16 == expected_raster_float16
+
+
+def test_astype_float32_to_float16_conversion_without_stretching():
+    raster_float16 = some_float32_raster.astype(np.float16, in_range=None, out_range=None)
+    expected_raster_float16 = GeoRaster2(
+        some_float32_array.astype(np.float16),
+        affine=some_float32_raster.affine, crs=some_float32_raster.crs, nodata=None)
+    assert raster_float16 == expected_raster_float16
+
+
+def test_astype_uint8_to_float32_conversion():
+    raster_float32 = some_raster.astype(np.float32, out_range=(0, 1))
+    expected_raster_float32 = GeoRaster2(image=np.ma.array(
+        np.array([
+            [0.0, 0.003921568859368563, 0.007843137718737125],
+            [0.0117647061124444, 0.01568627543747425, 1.0]
+        ], dtype=np.float32), mask=some_raster.image.mask),
+        affine=some_float32_raster.affine, crs=some_float32_raster.crs, nodata=None)
+    assert raster_float32 == expected_raster_float32
+
+
+def test_astype_uint8_to_float32_conversion_with_image_in_range():
```

Comment:
I recommend to check the influence of the no data mask on `in_range='image'`

Reply:
Such checking is performed:

```
# some_raster - raster from test
(Pdb) some_raster.image.data
array([[[0, 1, 2],
        [3, 4, 5]]], dtype=uint8)
(Pdb) some_raster.image.mask
array([[[False, False, False],
        [False, False,  True]]])

# expected raster after conversion
(Pdb) some_raster.astype(np.float32, in_range='image', out_range=(0, 1)).image
masked_array(
  data=[[[0.0, 0.25, 0.5],
         [0.75, 1.0, --]]],
  mask=[[[False, False, False],
         [False, False,  True]]],
  fill_value=999999.0,
  dtype=float32)

# manually passed in_range without taking in account mask:
(Pdb) some_raster.astype(np.float32, in_range=(0, 5), out_range=(0, 1)).image
masked_array(
  data=[[[0.0, 0.20000000298023224, 0.4000000059604645],
         [0.6000000238418579, 0.800000011920929, --]]],
  mask=[[[False, False, False],
         [False, False,  True]]],
  fill_value=999999.0,
  dtype=float32)

# manually passed in_range with taking in account mask
some_raster.astype(np.float32, in_range=(0, 4), out_range=(0, 1)).image
masked_array(
  data=[[[0.0, 0.25, 0.5],
         [0.75, 1.0, --]]],
  mask=[[[False, False, False],
         [False, False,  True]]],
  fill_value=999999.0,
  dtype=float32)
```

As you can see we are taking in accout mask in this test.